rest/route/host_routes.go

```diff
@@ -328,3 +335,276 @@ func (hph *hostPostHandler) Execute(ctx context.Context, sc data.Connector) (Res
+			StatusCode: http.StatusInternalServerError,
+			Message:    err.Error(),
+		}
+	}
+
+	return ResponseData{}, nil
+}
+
+func getHostExtendExpirationRouteManager(route string, version int) *RouteManager {
+	return &RouteManager{
+		Route:   route,
+		Version: version,
+		Methods: []MethodHandler{
+			{
+				PrefetchFunctions: []PrefetchFunc{PrefetchUser},
+				MethodType:        http.MethodPost,
+				Authenticator:     &RequireUserAuthenticator{},
+				RequestHandler:    &hostExtendExpirationHandler{},
+			},
+		},
+	}
+}
+
+type hostExtendExpirationHandler struct {
+	hostID   string
+	addHours time.Duration
+}
+
+func (h *hostExtendExpirationHandler) Handler() RequestHandler {
+	return &hostExtendExpirationHandler{}
+}
+
+func (h *hostExtendExpirationHandler) ParseAndValidate(ctx context.Context, r *http.Request) error {
+	hostModify := model.APISpawnHostModify{}
+	if err := util.ReadJSONInto(util.NewRequestReader(r), &hostModify); err != nil {
+		return err
+	}
+
+	var err error
+	h.hostID, err = validateHostID(mux.Vars(r)["host_id"])
+	if err != nil {
+		return err
+	}
+
+	addHours, err := strconv.Atoi(string(hostModify.AddHours))
+	if err != nil {
+		return &rest.APIError{
+			StatusCode: http.StatusBadRequest,
+			Message:    "expiration not a number",
+		}
+	}
+	h.addHours = time.Duration(addHours) * time.Hour
+
+	if h.addHours <= 0 {
+		return &rest.APIError{
+			StatusCode: http.StatusBadRequest,
+			Message:    "must add more than 0 hours to expiration",
+		}
+	}
+	if h.addHours > spawn.MaxExpirationDurationHours {
+		return &rest.APIError{
+			StatusCode: http.StatusBadRequest,
+			Message:    fmt.Sprintf("cannot add more than %s", spawn.MaxExpirationDurationHours.String()),
+		}
+	}
+
+	return nil
+}
+
+func (h *hostExtendExpirationHandler) Execute(ctx context.Context, sc data.Connector) (ResponseData, error) {
+	u := MustHaveUser(ctx)
+
+	host, err := fetchHostAndAuthorize(sc, h.hostID, u)
+	if err != nil {
+		return ResponseData{}, err
+	}
+	if host.Status == evergreen.HostTerminated {
+		return ResponseData{}, &rest.APIError{
+			StatusCode: http.StatusBadRequest,
+			Message:    "cannot extend expiration of a terminated host",
+		}
+	}
+
+	var newExp time.Time
+	newExp, err = spawn.MakeExtendedHostExpiration(host, h.addHours)
+	if err != nil {
+		return ResponseData{}, &rest.APIError{
+			StatusCode: http.StatusBadRequest,
+			Message:    err.Error(),
+		}
+	}
+
+	if err := sc.SetHostExpirationTime(host, newExp); err != nil {
+		return ResponseData{}, &rest.APIError{
+			StatusCode: http.StatusInternalServerError,
+			Message:    err.Error(),
+		}
+	}
+
+	return ResponseData{}, nil
+}
+
+func validateHostID(hostID string) (string, error) {
+	if strings.TrimSpace(hostID) == "" {
+		return "", &rest.APIError{
+			StatusCode: http.StatusBadRequest,
+			Message:    "missing/empty host id",
+		}
+	}
+
+	return hostID, nil
+}
+
+func fetchHostAndAuthorize(sc data.Connector, hostID string, user auth.User) (*host.Host, error) {
```

Comment:
why not have the data connector just take a host id and a user and have this logic be in the data package?